streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A Metered {@link KeyValueStore} wrapper that is used for recording operation metrics, and hence its
+ * inner KeyValueStore implementation do not need to provide its own metrics collecting functionality.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class MeteredKeyValueBytesStore<K, V> extends WrappedStateStore.AbstractStateStore implements KeyValueStore<K, V> {
+
+    private final Serde<K> keySerde;
+    private final Serde<V> valueSerde;
+    private StateSerdes<K, V> serdes;
+    private InnerMeteredKeyValueStore<K, Bytes, V, byte[]> innerMetered;
+
+    // always wrap the store with the metered store
+    public MeteredKeyValueBytesStore(final KeyValueStore<Bytes, byte[]> inner,
+                                     final String metricScope,
+                                     final Time time,
+                                     final Serde<K> keySerde,
+                                     final Serde<V> valueSerde) {
+        super(inner);
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        innerMetered = new InnerMeteredKeyValueStore<>(inner, metricScope, new InnerMeteredKeyValueStore.TypeConverter<K, Bytes, V, byte[]>() {
+            @Override
+            public Bytes innerKey(final K key) {
+                return Bytes.wrap(serdes.rawKey(key));
+            }
+
+            @Override
+            public byte[] innerValue(final V value) {
+                return serdes.rawValue(value);
```

Comment:
Not a comment: as we have concluded in https://issues.apache.org/jira/browse/KAFKA-4750 we should check for null here and skip. cc @evis.